File: worker/src/parsers/ss.ts
/**
 * ShadowSocks parser
 */

import { ProxyNode } from '../utils/types';
import { padBase64 } from '../services/subscription';
import { cleanQuery, decodeName } from '../utils/url';

export function parseSS(entry: string): ProxyNode {
  let body = entry.slice('ss://'.length);
  let name: string | undefined;

  const hashIndex = body.indexOf('#');
  if (hashIndex !== -1) {
    name = decodeName(body.slice(hashIndex));
    body = body.slice(0, hashIndex);
  }

  let query: Record<string, string> = {};
  const queryIndex = body.indexOf('?');
  if (queryIndex !== -1) {
    query = cleanQuery(body.slice(queryIndex + 1));
    body = body.slice(0, queryIndex);
  }

  let decodedBody = body;
  if (!body.includes('@')) {
    try {
      decodedBody = atob(padBase64(body));
    } catch {
      decodedBody = body;
    }
  }

  if (!decodedBody.includes('@') || !decodedBody.includes(':')) {
    throw new Error('invalid ss entry');
  }

  const atIndex = decodedBody.lastIndexOf('@');
  const userInfo = decodedBody.slice(0, atIndex);
  const hostInfo = decodedBody.slice(atIndex + 1);
  const colonIndex = userInfo.indexOf(':');
  if (colonIndex === -1) {
    throw new Error('invalid ss entry');
  }

  const method = userInfo.slice(0, colonIndex);
  const password = userInfo.slice(colonIndex + 1);
  const hostColonIndex = hostInfo.lastIndexOf(':');
  if (hostColonIndex === -1) {
    throw new Error('invalid ss entry');
  }

  const server = hostInfo.slice(0, hostColonIndex);
  const port = hostInfo.slice(hostColonIndex + 1);

  return {
    scheme: 'ss',
    name,
    server,
    port,
    cipher: method,
    password,
    query,
    raw: entry,
  };
}

File: src/sub_converter/parsers/ss.py
"""ShadowSocks (ss://) parser."""

from __future__ import annotations

import base64
import binascii
from typing import Any

from sub_converter.utils import clean_query, decode_name, pad_base64


def parse_ss(url: str) -> dict[str, Any]:
    body = url[len("ss://"):]
    name = None
    if "#" in body:
        body, fragment = body.split("#", 1)
        name = decode_name(fragment)

    query: dict[str, str] = {}
    if "?" in body:
        body, query_string = body.split("?", 1)
        query = clean_query(query_string)

    decoded_body = body
    if "@" not in body:
        maybe = pad_base64(body)
        try:
            decoded_body = base64.b64decode(maybe).decode("utf-8", errors="strict")
        except (binascii.Error, ValueError, UnicodeDecodeError):
            decoded_body = body

    if "@" not in decoded_body or ":" not in decoded_body:
        raise ValueError("invalid ss entry")

    userinfo, hostinfo = decoded_body.rsplit("@", 1)
    method, password = userinfo.split(":", 1)
    server, port_text = hostinfo.rsplit(":", 1)

    return {
        "scheme": "ss",
        "name": name,
        "cipher": method,
        "password": password,
        "server": server,
        "port": int(port_text),
        "query": query,
        "raw": url,
    }
